AppController/djinn.rb

```diff
@@ -5418,25 +5364,95 @@ def find_lowest_free_port(starting_port, ending_port=0, appid="")
   end
 
 
-  # Adds or removes AppServers within a node based on the number of requests
-  # that each application has received as well as the number of requests that
-  # are sitting in haproxy's queue, waiting to be served.
-  def scale_appservers_within_nodes
+  # Scale AppServers up/down for each application depending on the current
+  # queued requests and load of the application.
+  #
+  # Returns:
+  #   An Integer indicating the number of AppServers that we couldn't
+  #   start for lack of resources.
+  def scale_appservers
+    needed_appservers = 0
     @apps_loaded.each { |app_name|
       initialize_scaling_info_for_app(app_name)
 
       # Get the desired changes in the number of AppServers.
       delta_appservers = get_scaling_info_for_app(app_name)
       if delta_appservers > 0
         Djinn.log_debug("Considering scaling up app #{app_name}.")
-        try_to_scale_up(app_name, delta_appservers)
+        unless try_to_scale_up(app_name, delta_appservers)
+          needed_appservers += delta_appservers
+        end
       elsif delta_appservers < 0
         Djinn.log_debug("Considering scaling down app #{app_name}.")
         try_to_scale_down(app_name, delta_appservers.abs)
       else
         Djinn.log_debug("Not scaling app #{app_name} up or down right now.")
       end
     }
+
+    return needed_appservers
+  end
+
+
+  # Adds or removes AppServers and/or nodes to the deployment, depending
+  # on the statistics of the application and the loads of the various
+  # services.
+  def scale_deployment
+    needed_appservers = scale_appservers
+    if needed_appservers > 0
+      Djinn.log_debug("Need to start VMs for #{needed_appservers} more AppServers.")
+    end
+
+    # Here we cound the number of machines we need to spawn, and the roles
+    # we need.
+    vms_to_spawn = 0
```

Comment:
Hmm ... I'm not quite sure: those are the numbers of new VMs we need to create. vm_id doesn't seem to specify anything useful here: we don't have any ID, and actually we don' t have the VMs yet. Perhaps vm_to_create? or num_vm_to_spawn/create? Or vms_desired/required?